File: core/src/main/java/ir/logicfan/core/ui/base/BaseActivity.java
package ir.logicfan.core.ui.base;

import android.app.Activity;
import android.os.Bundle;
import android.view.MenuItem;
import androidx.annotation.IdRes;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import dagger.android.support.DaggerAppCompatActivity;
import ir.logicfan.core.data.pref.BasePrefManager;
import ir.logicfan.core.di.qulifier.ActivityFragmentManager;
import ir.logicfan.core.util.LocaleManager;

import javax.inject.Inject;

/**
 * Created by Mahdi on 12/16/2017.
 * All of our activities extends this class to inherit top level functionality
 */

public abstract class BaseActivity extends DaggerAppCompatActivity {

    @Inject
    BasePrefManager prefManager;

    @Inject
    @ActivityFragmentManager
    protected FragmentManager fragmentManager;

    private Unbinder butterKnifeUnbinder;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        LocaleManager.setLocale(this, prefManager.getSettingsPrefLangList());
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
            return true;
        } else {
            return super.onOptionsItemSelected(item);
        }
    }

    protected void replaceFragmentWithFadeAnimation(@IdRes int container, Fragment fragment, boolean addToBackStack) {
        FragmentTransaction ft = fragmentManager.beginTransaction();
        ft.replace(container, fragment);
        if (addToBackStack) {
            ft.addToBackStack(null); // argument name is optional
        }
        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);  // add animation to replacement process
        ft.commit();
    }

    protected void setButterKnifeUnbinder(Activity activity) {
        butterKnifeUnbinder = ButterKnife.bind(activity);
    }

    @Override
    protected void onDestroy() {
        if (butterKnifeUnbinder != null) {
            butterKnifeUnbinder.unbind();
        }
        super.onDestroy();
    }
}

File: core/src/main/java/ir/logicfan/core/ui/base/BaseActivity.kt
package ir.logicfan.core.ui.base

import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.activity.ComponentActivity
import androidx.annotation.CallSuper
import androidx.annotation.MainThread
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelLazy
import androidx.lifecycle.ViewModelProvider
import dagger.android.AndroidInjection
import dagger.android.AndroidInjector
import dagger.android.DispatchingAndroidInjector
import dagger.android.HasAndroidInjector
import ir.logicfan.core.R
import ir.logicfan.core.data.error.DataException
import ir.logicfan.core.data.sharedpreferences.BaseSharedPreferences
import ir.logicfan.core.databinding.CoreViewNetworkUnavailableBinding
import ir.logicfan.core.ui.feature.main.NetworkConnectivityReceiver
import ir.logicfan.core.ui.feature.main.NetworkConnectivityViewModel
import ir.logicfan.core.ui.util.LocaleUtils
import javax.inject.Inject

/**
 * All of our activities extends this class to inherit top level functionality
 */
abstract class BaseActivity : AppCompatActivity(), HasAndroidInjector, DataTerminalErrorListener {

    @Inject
    lateinit var androidInjector: DispatchingAndroidInjector<Any>
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    @Inject
    lateinit var networkConnectivityReceiver: NetworkConnectivityReceiver
    @Inject
    lateinit var baseSharedPreferences: BaseSharedPreferences

    private val networkConnectivityViewModel: NetworkConnectivityViewModel by viewModels()
    open fun attachBaseViewModel(): List<BaseViewModel>? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        AndroidInjection.inject(this) // inject dagger
        super.onCreate(savedInstanceState)
        LocaleUtils.setLocale(this, baseSharedPreferences.localeSetting) // setup locale
        networkConnectivityReceiver.observe(this, Observer {
            if (it) {
                // network connection is active now
                networkConnectivityViewModel.onNetworkAvailabilityChange(true)
            }
        })

        // handle terminal errors
        val baseViewModels = attachBaseViewModel()
        baseViewModels?.forEach { viewModel ->
            viewModel.errorState.observe(this, Observer {
                onDataTerminalError(it)
            })
        }
        networkConnectivityViewModel.networkBecomesAvailable.observe(this, Observer {
            baseViewModels?.forEach { viewModel ->
                viewModel.onNetworkBecomesAvailable()
            }
        })

        networkConnectivityViewModel.alertPanelVisibility.observe(this, Observer {
            val container = attachNetworkAvailabilityContainer()
            if (container != null && it) {
                val viewBinding = DataBindingUtil.inflate<CoreViewNetworkUnavailableBinding>(
                    layoutInflater, attachNetworkAvailabilityLayoutRes(), container, true
                )
                viewBinding.onCloseClickListener = View.OnClickListener { networkConnectivityViewModel.onCloseClick() }
            } else {
                container?.removeAllViews()
            }
        })
    }

    open fun attachNetworkAvailabilityContainer(): ViewGroup? = null

    protected fun attachNetworkAvailabilityLayoutRes(): Int = R.layout.core_view_network_unavailable

    @CallSuper
    override fun onDataTerminalError(throwable: Throwable) {
        if (throwable  is DataException.Terminal.Offline) {
            networkConnectivityViewModel.onNetworkAvailabilityChange(false)
        }
    }

    @MainThread
    inline fun <reified VM : ViewModel> ComponentActivity.viewModels(
        noinline factoryProducer: (() -> ViewModelProvider.Factory)? = { viewModelFactory }
    ): Lazy<VM> {
        val factoryPromise = factoryProducer ?: {
            val application = application ?: throw IllegalArgumentException(
                "ViewModel can be accessed only when Activity is attached"
            )
            ViewModelProvider.AndroidViewModelFactory.getInstance(application)
        }

        return ViewModelLazy(VM::class, { viewModelStore }, factoryPromise)
    }

    override fun androidInjector(): AndroidInjector<Any> = androidInjector
}
